File: 45-jump-game-ii/45-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        
        if(n == 1) return 0;
        
        int jumps = 1;
        int maxReach = nums[0];
        int winEnd = nums[0];
        
        for(int i = 1; i < n - 1; ++i) {
            maxReach = max(maxReach, i + nums[i]);
            if(i == winEnd) {
                ++jumps;
                winEnd = maxReach;
            }
        }
        return jumps;
    }
};

File: 45-jump-game-ii/45-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        
        if n == 1: return 0
        
        jumps = 1
        maxReach = nums[0]
        winEnd = nums[0]
        
        for i in range(1, n - 1):
            maxReach = max(maxReach, i + nums[i])
            if i == winEnd:
                jumps += 1
                winEnd = maxReach
        
        return jumps